pkg/initializer/initializer.go

```diff
@@ -45,23 +45,29 @@ import (
 //   - Try to perform an Etcd data restoration from the latest snapshot.
 //   - No snapshots are available, start etcd as a fresh installation.
 func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int64) error {
+	logger := e.Logger.WithField("actor", "initializer")
 	metrics.CurrentClusterSize.With(prometheus.Labels{}).Set(float64(e.Validator.OriginalClusterSize))
 	start := time.Now()
 	isEtcdMemberPresent := false
 	ctx := context.Background()
 	var err error
 
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Fatalf("failed to create clientset, %v", err)
+	}
+
 	//Etcd cluster scale-up case
-	if miscellaneous.IsMultiNode(e.Logger.WithField("actor", "initializer")) {
+	if miscellaneous.IsMultiNode(logger) {
 		m := member.NewMemberControl(e.Config.EtcdConnectionConfig)
-		isEtcdMemberPresent, err = m.IsMemberInCluster(ctx)
-		if !isEtcdMemberPresent && err == nil {
-			retry.OnError(retry.DefaultBackoff, func(err error) bool {
-				return err != nil
-			}, func() error {
+		isScaleup, err := m.IsClusterScaledup(ctx, clientSet)
+		if isScaleup && err == nil {
```

Comment:
if a call to `IsClusterScaledup` returns an error then you currently assume that it is not a scaleup case and completely skip `AddMemberAsLeaner`.

Reply:
>I found that if IsClusterScaledup returns an error we have to assume that it is not a scaleup case else bootstrap case 0->3 won't be possible.

There is always a possibility that an error is returned even in the case of a scale-up when calling `IsMemberInCluster` - therefore assuming that if there is an error therefore it is not a scale up case but a bootstrap (0-3) case is IMHO incorrect.

Reply:
Its very difficult to distinguish between the errors here
For instance, `IsMemberInCluster` internally makes a call to `ProbeEtcd`. This could fail because we're in the bootstrapping case and etcds are not running, or due to some network issue/other issue. It's very difficult to deterministically differentiate between these two right now